File: dsa-csharp/solutions/LinkedList/Medium/DesignLinkedList.cs
// 707. Design Linked List
// https://leetcode.com/problems/design-linked-list/
//
// Implement a singly linked list with get, addAtHead, addAtTail, addAtIndex, deleteAtIndex.
//
// APPROACH: Standard singly linked list with a head pointer.
// Each node holds a value and a pointer to the next node.
//
// WHY no dummy node here: We track `Head` directly. This means we need to
// handle edge cases (empty list, insert at head) separately, but it's straightforward.
//
// TIME: Get/Add/Delete are all O(n) in worst case because we may traverse the entire list.
// SPACE: O(n) for storing n nodes.

public class MyLinkedList
{
    private Node Head;

    public MyLinkedList()
    {
        Head = null; // Start with an empty list
    }

    // Get the value at the given index (0-based). Return -1 if index is invalid.
    public int Get(int index)
    {
        if (Head == null) return -1;

        Node current = Head;

        // Move forward `index` times to reach the target node
        for (int i = 0; i < index; i++)
        {
            if (current == null) return -1; // index out of bounds
            current = current.next;
        }

        // If current is null, index was too large
        return current == null ? -1 : current.value;
    }

    // Insert a new node at the very beginning of the list
    public void AddAtHead(int val)
    {
        Node node = new Node(val);
        node.next = Head; // New node points to the old head
        Head = node;      // New node becomes the head
    }

    // Insert a new node at the end of the list
    public void AddAtTail(int val)
    {
        Node node = new Node(val);

        // If list is empty, the new node IS the head
        if (Head == null) { Head = new Node(val); return; }

        // Traverse to the last node
        Node current = Head;
        while (current.next != null)
        {
            current = current.next;
        }
        // Attach new node at the end
        current.next = node;
    }

    // Insert a new node BEFORE the node at `index`.
    // If index == 0, insert at head. If index > length, don't insert.
    public void AddAtIndex(int index, int val)
    {
        // Special case: inserting at position 0 means inserting at head
        if (index == 0)
        {
            AddAtHead(val);
            return;
        }

        Node current = Head;

        // We need to stop at the node BEFORE the target index (index - 1)
        // so we can insert after it
        for (int i = 0; i < index - 1; i++)
        {
            if (current == null) return; // index out of bounds
            current = current.next;
        }

        if (current == null) return; // index out of bounds

        // Insert: new node points to current's next, current points to new node
        Node node = new Node(val);
        node.next = current.next;
        current.next = node;
    }

    // Delete the node at the given index
    public void DeleteAtIndex(int index)
    {
        if (Head == null) return;

        // Special case: deleting head — just move head to next
        if (index == 0)
        {
            Head = Head?.next;
            return;
        }

        Node current = Head;

        // Navigate to the node BEFORE the one we want to delete
        for (int i = 0; i < index - 1; i++) 
        {
            if (current.next == null) return; // index out of bounds
            current = current.next;
        }

        if (current.next == null) return; // nothing to delete

        // Skip over the target node: prev.next = target.next
        current.next = current.next.next;
    }
}


public class Node
{
    public Node next;
    public int value;

    public Node(int value)
    {
        this.value = value;
        this.next = null;
    }
}
/**
 * Your MyLinkedList object will be instantiated and called as such:
 * MyLinkedList obj = new MyLinkedList();
 * int param_1 = obj.Get(index);
 * obj.AddAtHead(val);
 * obj.AddAtTail(val);
 * obj.AddAtIndex(index,val);
 * obj.DeleteAtIndex(index);
 */

File: dsa-python/solutions/linked_list/medium/design_linked_list.py
from solutions import ListNode


# 707. Design Linked List
# https://leetcode.com/problems/design-linked-list/
#
# Implement a singly linked list with get, add_at_head, add_at_tail, add_at_index, delete_at_index.
#
# APPROACH: Standard singly linked list with a head pointer.
# We reuse the ListNode class (val, next) from __init__.py.
#
# WHY no dummy node: We track `head` directly. This means we handle edge cases
# (empty list, insert at head) as special cases.
#
# TIME: All operations are O(n) worst case (must traverse to find the position).
# SPACE: O(n) for storing n nodes.

class MyLinkedList:
    def __init__(self):
        self.head = None  # Start with an empty list

    def get(self, index: int) -> int:
        """Get value at index (0-based). Return -1 if index is invalid."""
        curr = self.head
        # Move forward `index` times to reach the target
        for _ in range(index):
            if curr is None:
                return -1  # index out of bounds
            curr = curr.next
        # If we landed on a valid node, return its value
        return curr.val if curr else -1

    def add_at_head(self, val: int) -> None:
        """Insert a new node at the beginning of the list."""
        new_node = ListNode(val)
        new_node.next = self.head  # New node points to old head
        self.head = new_node       # New node becomes the head

    def add_at_tail(self, val: int) -> None:
        """Insert a new node at the end of the list."""
        newNode = ListNode(val)

        # If list is empty, new node IS the head
        if(self.head is None):
            self.head = newNode
            return
        
        # Traverse to the last node
        curr_node = self.head
        while(curr_node.next):
            curr_node = curr_node.next
        
        # Attach new node at the end
        curr_node.next = newNode

    def add_at_index(self, index: int, val: int) -> None:
        """Insert a new node BEFORE the node at `index`.
        If index == 0, insert at head. If index > length, don't insert."""
        newNode = ListNode(val)

        # Special case: inserting at position 0 = inserting at head
        if(index == 0):
            self.add_at_head(val)
            return
        
        # Navigate to the node BEFORE the target index (index - 1)
        # so we can insert after it
        i = 0
        curr = self.head
        while(i < index - 1):
            if(curr is None):
                return  # index out of bounds
            curr = curr.next
            i = i + 1
        
        if(curr is None):
            return  # index out of bounds

        # Insert: new node points to curr's next, curr points to new node
        newNode.next = curr.next
        curr.next = newNode

    def delete_at_index(self, index: int) -> None:
        """Delete the node at the given index."""
        # Special case: deleting head — just move head forward
        if(index == 0):
            self.head = self.head.next
            return
        
        curr = self.head
        
        # Navigate to the node BEFORE the one we want to delete
        for _ in range(index - 1):
            if(curr is None):
                return  # index out of bounds
            curr = curr.next
        
        if(curr == None):
            return  # nothing to delete

        # Skip over the target: prev.next = target.next
        curr.next = curr.next.next
